File: java/CoursioApi.java
package CoursioApi;

import java.net.*;
import java.io.*;
import java.sql.Timestamp;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public class CoursioApi
{
	protected HttpURLConnection connection;

	protected String salt;
	protected String publicKey;
	protected String privateKey;
	protected String baseUrl = "https://api.coursio.com/v1/";
	
    public CoursioApi(String publicKey, String privateKey, String salt) throws Exception
    {
		if (publicKey == null || privateKey == null || salt == null)
		{
			throw new Exception ("Both keys and salt are required.");
		}

		this.salt = salt;
		this.publicKey = publicKey;
		this.privateKey = privateKey;
    }
    
    public String Get(String endpoint, int objectId) throws Exception
	{
		if (objectId != 0)
		{
			endpoint += "/" + Integer.toString(objectId);
		}

		Prepare (endpoint);
		connection.setRequestMethod("GET");
		
		return ReadResponse ();
	}

	public String Post(String endpoint, String jsonString) throws Exception
	{
		Prepare (endpoint);
		connection.setRequestMethod("POST");
		
        OutputStreamWriter osw = new OutputStreamWriter(connection.getOutputStream());
        osw.write(jsonString);
        osw.flush();
        osw.close();

		return ReadResponse ();
	}

	public String Put(String endpoint, int objectId, String jsonString) throws Exception
	{
		if (objectId == 0)
		{
			throw new Exception ("No object ID specified");
		}
		endpoint += "/" + Integer.toString(objectId);

		Prepare (endpoint);
		connection.setRequestMethod("PUT");

        OutputStreamWriter osw = new OutputStreamWriter(connection.getOutputStream());
        osw.write(jsonString);
        osw.flush();
        osw.close();

		return ReadResponse ();
	}

	public String Delete(String endpoint, int objectId) throws Exception
	{
		if (objectId == 0)
		{
			throw new Exception ("No object ID specified");
		}
		endpoint += "/" + Integer.toString(objectId);

		Prepare (endpoint);
		connection.setRequestMethod("DELETE");

		return ReadResponse ();
	}

	protected void Prepare(String endpoint) throws Exception
	{
		if (endpoint == null || endpoint.length() == 0)
		{
			throw new Exception ("No endpoint specified");
		}

		// setup connection to endpoint
        URL obj = new URL(baseUrl + endpoint);
        connection = (HttpURLConnection) obj.openConnection();
        
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setDoOutput(true);

		// compute HMAC
        
        // get an hmac_sha1 key from the raw key bytes
        SecretKeySpec signingKey = new SecretKeySpec(privateKey.getBytes(), "HmacSHA1");

        // get an hmac_sha1 Mac instance and initialize with the signing key
        Mac mac = Mac.getInstance("HmacSHA1");
        mac.init(signingKey);
        
        // generate timestamp
        java.util.Date date = new java.util.Date();
        Timestamp timestamp = new Timestamp(date.getTime());

        String signature = publicKey + timestamp.toString() + salt;
        
        // compute the hmac on input data bytes
        byte[] bytes = mac.doFinal(signature.getBytes());
        StringBuilder hash = new StringBuilder();
        
        for (byte b : bytes)
        {
        	hash.append(String.format("%02x", b));
        }
        
        connection.setRequestProperty ("X-Coursio-apikey", publicKey);
        connection.setRequestProperty ("X-Coursio-time", timestamp.toString());
        connection.setRequestProperty ("X-Coursio-random", salt);
        connection.setRequestProperty ("X-Coursio-hmac", hash.toString());
	}

	protected String ReadResponse() throws IOException
	{
		String result = "";
		
        // Get the response
        BufferedReader rd = new BufferedReader(new InputStreamReader(connection.getInputStream()));
        String line;
        while ((line = rd.readLine()) != null)
        {
        	result += line;
        }
        rd.close();
        
        return result;
	}
}

File: python/CoursioApi.py
import time
from hashlib import sha1
import hmac
import binascii
import json
import httplib2 as http

try:
    from urlparse import urlparse
except ImportError:
    from urllib.parse import urlparse

class CoursioApi:

    # initiate the class
    def __init__(self, publicKey, privateKey, salt):
        self.publicKey = publicKey
        self.privateKey = privateKey
        self.salt = salt
        self.baseUrl = 'http://api.coursio.dev'
        self.request = ''
        self.headers = {}
        self.target = ''
        self.method = 'GET'
        self.h = ''
        self.body = ''

    # prepare an endpoint and caluclate HMAC
    def prepare(self, endpoint):
        if (endpoint is None):
            raise Exception('No endpoint specified')

        # calculate microtime, first to int then to string
        microtime = str(int(round(time.time() * 1000)))

        # concatenate the raw string
        rawString = self.publicKey + microtime + self.salt

        hashed = hmac.new(self.privateKey, rawString, sha1)

        # get hmac from calculated hash
        hmacString = binascii.b2a_base64(hashed.digest())[:-1]

        # setup headers
        self.headers = {
            'X-Coursio-apikey': self.publicKey,
            'X-Coursio-time': microtime,
            'X-Coursio-random': self.salt,
            'X-Coursio-hmac': hmacString
        }

        # concatenate target
        self.target = urlparse(self.baseUrl+endpoint)

        # setup http
        self.h = http.Http()

    # get method
    def get(self, endpoint, object_id = 0):
        # cast object_id to int
        object_id = int(object_id)

        # if object_id exists, add to endpoint
        if (object_id != 0):
            endpoint = endpoint + '/' + str(object_id)

        # prepare for call
        self.prepare(endpoint)

        # set request-method
        self.method = 'GET'

        return self.response()

    def response(self):

        # build request and store response and concent in variables
        response, content = self.h.request(
            self.target.geturl(),
            self.method,
            self.body,
            self.headers
        )

        # return json from server
        return content